Removed unnecessary term swap while computing Mul in parse_expr.

dsaw opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes Mul issue of #21119


#### Brief description of what is fixed or changed
A swapping happens in `parse_expr`(evaluate=False) when operation is `Mul` that was changing the order, is removed since it is since unnecessary now.

#### Other comments
Note the #10560 for which the old swapping #10774 fix was done might be need to reopened since some of the nested terms `('9.2*(4.1+(-4)*3)')` aren't working, even with the old fix.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* parsing
  * Fix to maintain the order of Mul terms in parse_expr when evaluate=False
<!-- END RELEASE NOTES -->

dsaw committed:
Removed unnecessary term swap while computing Mul in parse_expr.

oscarbenjamin commented on sympy/parsing/sympy_parser.py:
```diff
@@ -1061,8 +1061,7 @@ def visit_BinOp(self, node):
             right = self.visit(node.right)
             left = self.visit(node.left)
-            if isinstance(node.left, ast.UnaryOp) and (isinstance(node.right, ast.UnaryOp) == 0) and sympy_class in ('Mul',):
-                left, right = right, left
```
I'm not sure I understand why this happens. Tracing through the blame this seems to come from https://github.com/sympy/sympy/pull/10774

@smichr do you have any idea why these are swapped?

smichr replied:
Must have been a work around for something that has since been fixed. I no longer recall any specifics other than what was recorded in the PR that did this.

oscarbenjamin replied:
Actually I think maybe the swap here is correct and shouldn't be removed.

The linked issue is to do with the printer not the parser: https://github.com/sympy/sympy/issues/21119#issuecomment-809534656

smichr replied:
But with this PR the following regresses:

in master
```python
>>> sympify('-10/5', evaluate = False)
1/5*(-10)
>>> _.args
(1/5, -10)
```
 now gives the incorrect result,
```
>>> sympify('-10/5', evaluate = False).args
(1/(-10), 5)
```

smichr replied:
And I am not sure what the OP comment means in light of the following correct parse in master:
```
>>> S('9.2*(4.1+(-4)*3)',evaluate=0)
9.2*(4.1 + (-4)*3)
```

smichr replied:
> linked issue is to do with the printer not the parser

I think there is a parser issue:
```
>>> S('3*4',evaluate=0).args
(3, 4)
>>> S('3*-4',evaluate=0).args
(3, -4)
>>> S('-3*4',evaluate=0).args
<reversal code traversed>
(4, -3)
>>> S('-3*-4',evaluate=0).args
(-3, -4)

>>> S('1/2',evaluate=0).args
(1, 1/2)
>>> S('1/-2',evaluate=0).args
(1, 1/(-2))
>>> S('-1/2',evaluate=0).args
<reversal code traversed>
(1/2, -1)
>>> S('-1/-2',evaluate=0).args
(1/(-2), -1)
```

dsaw replied:
@smichr Right, missed the `Div` case in here, will update.
I think we can move that swap in the `ast.Div` if statement itself, since doesn't seem to be needed for `ast.Mul` operator  like this.
```
--- a/sympy/parsing/sympy_parser.py
+++ b/sympy/parsing/sympy_parser.py
@@ -1062,7 +1062,12 @@ def visit_BinOp(self, node):
             sympy_class = self.operators[node.op.__class__]
             right = self.visit(node.right)
             left = self.visit(node.left)
-            if isinstance(node.left, ast.UnaryOp) and (isinstance(node.right, ast.UnaryOp) == 0) and sympy_class in ('Mul',):
-                left, right = right, left
+            rev = False
             if isinstance(node.op, ast.Sub):
                 right = ast.Call(
                     func=ast.Name(id='Mul', ctx=ast.Load()),
@@ -1073,8 +1078,8 @@ def visit_BinOp(self, node):
                 )
             if isinstance(node.op, ast.Div):
                 if isinstance(node.left, ast.UnaryOp):
-                    if isinstance(node.right,ast.UnaryOp):
-                        left, right = right, left
+                    left, right = right, left
+                    rev = True
                     left = ast.Call(
                     func=ast.Name(id='Pow', ctx=ast.Load()),
                     args=[left, ast.UnaryOp(op=ast.USub(), operand=ast.Num(1))],
@@ -1091,6 +1096,8 @@ def visit_BinOp(self, node):
                     kwargs=None
                 )

+            if rev:  # undo reversal
+                left, right = right, left
```

smichr commented:
This diff over master appears to resolve the issue:
```diff
diff --git a/sympy/parsing/sympy_parser.py b/sympy/parsing/sympy_parser.py
index bd2e020c53..d26e3b336e 100644
--- a/sympy/parsing/sympy_parser.py
+++ b/sympy/parsing/sympy_parser.py
@@ -1062,8 +1062,13 @@ def visit_BinOp(self, node):
             sympy_class = self.operators[node.op.__class__]
             right = self.visit(node.right)
             left = self.visit(node.left)
-            if isinstance(node.left, ast.UnaryOp) and (isinstance(node.right, ast.UnaryOp) == 0) and sympy_class in ('Mul',):
+            if (isinstance(node.left, ast.UnaryOp) and
+                    not isinstance(node.right, ast.UnaryOp) and
+                    sympy_class in ('Mul',)):
                 left, right = right, left
+                rev = True
+            else:
+                rev = False  # there was no reversal
             if isinstance(node.op, ast.Sub):
                 right = ast.Call(
                     func=ast.Name(id='Mul', ctx=ast.Load()),
@@ -1072,7 +1077,7 @@ def visit_BinOp(self, node):
                     starargs=None,
                     kwargs=None
                 )
-            if isinstance(node.op, ast.Div):
+            elif isinstance(node.op, ast.Div):
                 if isinstance(node.left, ast.UnaryOp):
                     if isinstance(node.right,ast.UnaryOp):
                         left, right = right, left
@@ -1091,7 +1096,8 @@ def visit_BinOp(self, node):
                     starargs=None,
                     kwargs=None
                 )
-
+            if rev:  # undo reversal
+                left, right = right, left
             new_node = ast.Call(
                 func=ast.Name(id=sympy_class, ctx=ast.Load()),
                 args=[left, right],
diff --git a/sympy/parsing/tests/test_sympy_parser.py b/sympy/parsing/tests/test_sympy_parser.py
index e019f4dce2..25c7e9e91b 100644
--- a/sympy/parsing/tests/test_sympy_parser.py
+++ b/sympy/parsing/tests/test_sympy_parser.py
@@ -158,9 +158,10 @@ def test_issue_7663():
     assert parse_expr(e, evaluate=0) == parse_expr(e, evaluate=False)
     assert parse_expr(e, evaluate=0).equals(2*(x+1))
 
-def test_issue_10560():
+
+def test_issue_10560_21119():
     inputs = {
-        '4*-3' : '(-3)*4',
+        '4*-3' : '4*(-3)',
         '-4*3' : '(-4)*3',
     }
     for text, result in inputs.items():

```

dsaw committed:
Restored swapping term behaviour for ast.Div nodes for evaluate=False instead of old condition which covered ast.Mul also. Added a reverse flag that it will swap to initial order if reversal happens.

Added numerator and denominator factor code to str printer that will separate into n1*.../ d1*d2... format (using @smichr example).

smichr commented:
maybe this will work?
```diff
diff --git a/sympy/printing/str.py b/sympy/printing/str.py
index 2b41fb706d..dc21014228 100644
--- a/sympy/printing/str.py
+++ b/sympy/printing/str.py
@@ -259,15 +259,19 @@ def _print_Mul(self, expr):
                 for a in args[1:]):
             from sympy.utilities.iterables import sift
             d, n = sift(args, lambda x:
-                isinstance(x, Pow) and x.exp == -1,
+                isinstance(x, Pow) and x.exp.is_Integer and bool(x.exp < 0),
                 binary=True)
-            d = [1/i for i in d]
+            d = [Pow(i.base, -i.exp, evaluate=False) if i.exp != -1 else 1/i for i in d]
             nfactors = [self.parenthesize(a, prec, strict=False)
                 for a in n]
             dfactors = [self.parenthesize(a, prec, strict=False)
                 for a in d]
             n = '*'.join(nfactors)
+            if n.startswith('(-1)*'):
+                n = '-' + n[5:]
             d = '*'.join(dfactors)
+            if d.startswith('(-1)*'):
+                d = '-' + d[5:]
             if len(dfactors) > 1:
                 return '%s/(%s)' % (n, d)
             elif dfactors:
```

smichr commented:
This passes doctests:
```diff
diff --git a/sympy/printing/str.py b/sympy/printing/str.py
index 9227e472ef..48eb036a77 100644
--- a/sympy/printing/str.py
+++ b/sympy/printing/str.py
@@ -11,7 +11,7 @@
 
 from mpmath.libmp import prec_to_dps, to_str as mlib_to_str
 
-from sympy.utilities import default_sort_key
+from sympy.utilities import default_sort_key, sift
 
 
 class StrPrinter(Printer):
@@ -254,26 +254,32 @@ def _print_Mul(self, expr):
         # etc so we display in a straight-forward form that fully preserves all
         # args and their order.
         args = expr.args
-        if args[0] is S.One or any(isinstance(arg, Number) or isinstance(1/arg, Number) for arg in args[1:]):
-            from sympy.utilities.iterables import sift
-            d, n  = sift(args, lambda x:
-                         isinstance(x,Pow) and x.exp == -1,
-                         binary=True
-                         )
-            d = [1/i for i in d]
-            nfactors = [self.parenthesize(arg, prec, strict=False) for arg in n]
-            dfactors = [self.parenthesize(arg, prec, strict=False) for arg in d]
-            numerator_string = '*'.join(nfactors)
-            denominator_string = '*'.join(dfactors)
+        if args[0] is S.One or any(
+                isinstance(a, Number) or
+                a.is_Pow and all(ai.is_Integer for ai in a.args)
+                for a in args[1:]):
+            d, n = sift(args, lambda x:
+                isinstance(x, Pow) and bool(x.exp.as_coeff_Mul()[0] < 0),
+                binary=True)
+            for i, di in enumerate(d):
+                if di.exp.is_Number:
+                    e = -di.exp
+                else:
+                    dargs = list(di.exp.args)
+                    dargs[0] = -dargs[0]
+                    e = Mul._from_args(dargs)
+                d[i] = Pow(di.base, e, evaluate=False) if e - 1 else di.base
+            nfactors = [self.parenthesize(a, prec, strict=False)
+                for a in n]
+            dfactors = [self.parenthesize(a, prec, strict=False)
+                for a in d]
+            n = '*'.join(nfactors)
+            d = '*'.join(dfactors)
             if len(dfactors) > 1:
-                return '%s/(%s)' % (numerator_string, denominator_string)
+                return '%s/(%s)' % (n, d)
             elif dfactors:
-                return '%s/%s' % (numerator_string, denominator_string)
-            else:
-                return numerator_string
-
-            factors = [self.parenthesize(a, prec, strict=False) for a in args]
-            return '*'.join(factors)
+                return '%s/%s' % (n, d)
+            return n
 
         c, e = expr.as_coeff_Mul()
         if c < 0:
diff --git a/sympy/printing/tests/test_str.py b/sympy/printing/tests/test_str.py
index 2b20c68725..69ff056640 100644
--- a/sympy/printing/tests/test_str.py
+++ b/sympy/printing/tests/test_str.py
@@ -1006,16 +1006,21 @@ def test_str_special_matrices():
     assert str(ZeroMatrix(2, 2)) == '0'
     assert str(OneMatrix(2, 2)) == '1'
 
+
 def test_issue_14567():
     assert factorial(Sum(-1, (x, 0, 0))) + y  # doesn't raise an error
 
+
 def test_issue_21119():
-    from sympy import sympify
-    assert str(sympify('4/2', evaluate=False)) == '4/2'
-    assert str(sympify('4/-2', evaluate=False)) == '4/(-2)'
-    assert str(sympify('-4/2', evaluate=False)) == '(-4)/2'
-    assert str(sympify('-4/-2', evaluate=False)) == '(-4)/(-2)'
-    assert str(sympify('4/2/1', evaluate=False)) == '4/(2*1)'
+    ss = lambda x: str(S(x, evaluate=False))
+    assert ss('4/2') == '4/2'
+    assert ss('4/-2') == '4/(-2)'
+    assert ss('-4/2') == '(-4)/2'
+    assert ss('-4/-2') == '(-4)/(-2)'
+    assert ss('4/2/1') == '4/(2*1)'
+    assert ss('2*3*4**(-2*3)') == '2*3/4**(2*3)'
+    assert ss('2*3*1*4**(-2*3)') == '2*3*1/4**(2*3)'
+
 
 def test_Str():
     from sympy.core.symbol import Str
```
